[dtensor] rewrite redistribute algorithm for multi-dim mesh

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132914
* #132683
* #132682
* __->__ #131210

As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

should also fix https://github.com/pytorch/pytorch/issues/132751

cc @XilunWu @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wanchaol committed:
[dtensor] rewrite redistribute algorithm

As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/_redistribute.py:
```diff
@@ -97,43 +86,131 @@ def _gen_transform_infos(
-                )
-            )
-        else:
+    if src_spec.num_shards > 1:
```
Can be replaced with `src_spec.num_shards_map[dim] > 1` once #130760 is merged

wanchaol committed:
Update on "[dtensor] rewrite redistribute algorithm for multi-dim mesh"


As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] rewrite redistribute algorithm for multi-dim mesh"


As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] rewrite redistribute algorithm for multi-dim mesh"


As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] rewrite redistribute algorithm for multi-dim mesh"


As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] rewrite redistribute algorithm for multi-dim mesh"


As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

tianyu-l reviewed (commented):
Left some nit comments on languages. I'm stuck at understanding `reshardable_from_src_to_dst`. I might need to discuss in person to understand more details before I can do further review...

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60132518).

wanchaol committed:
Update on "[dtensor] rewrite redistribute algorithm for multi-dim mesh"


As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D60132518](https://our.internmc.facebook.com/intern/diff/D60132518)

[ghstack-poisoned]

tianyu-l commented on torch/distributed/_tensor/_redistribute.py:
```diff
@@ -97,43 +86,131 @@ def _gen_transform_infos(
+                    mesh_dim=current_mesh_idx,
                     src_dst_placements=(src, dst),
-                    logical_shape=mesh_dims_to_logical_shape[i],
+                    logical_shape=mesh_dims_to_logical_shape[current_mesh_idx],
```
When doing right to left pass, `mesh_dims_to_logical_shape` is not modified when `src` is shard and `dst` is replicate. Would this become an issue if we go from left to right, since `mesh_dims_to_logical_shape` would become obsolete?

wanchaol replied:
I don't think it would have any effect here, the reason is that `mesh_dims_to_logical_shape` is mainly used for replication (shard -> replicate) purpose where we need to know the logical shape when it's uneven sharded, so that it would be used when going from right to left, but not left to right.

tianyu-l reviewed (commented):
For 2D I think the algorithm is robust.

I have question on 3+ D? For example, suppose `src_placements=[S(0), S(1), S(0)]`, `dst_placements=[S(1), S(0), S(0)]`.
When doing right to left we get:
`[S(0), S(1), S(0)]` -> `[S(0), S(1), S(0)]` (`mesh_dim` = 2)
`[S(0), S(1), S(0)]` -> `[S(0), R, S(0)]` (`mesh_dim` = 1)
`[S(0), R, S(0)]` -> `[R, R, S(0)]` (`mesh_dim` = 0)
But here the S(0) on `mesh_dim=2` actually becomes a strided sharding implicitly. Maybe this implicit strided sharding is OK?

Even if it's OK, I think we can remove the check `if src.is_shard()` in `reshardable_from_src_to_dst` and get the following from right to left:
`[S(0), S(1), S(0)]` -> `[S(0), S(1), S(0)]` (`mesh_dim` = 2)
`[S(0), S(1), S(0)]` -> `[S(0), R, S(0)]` (`mesh_dim` = 1)
`[S(0), R, S(0)]` -> `[S(1), R, S(0)]` (`mesh_dim` = 0)

Originally I thought the expected order from right to left should be:
`[S(0), S(1), S(0)]` -> `[S(0), S(1), R]` (`mesh_dim` = 2)
`[S(0), S(1), R]` -> `[S(0), R, R]` (`mesh_dim` = 1)
`[S(0), R, R]` -> `[S(1), R, R]` (`mesh_dim` = 0)

wanchaol commented on torch/distributed/_tensor/_redistribute.py:
```diff
@@ -97,43 +86,131 @@ def _gen_transform_infos(
+
+    # For src that is Shard, check if it's the inner most sharding
+    # if not then it's not reshardable, else we check the dst sharding
+    if src.is_shard():
+        for i in reversed(range(len(src_placements))):
+            if src_placements[i].is_shard(src.dim):
+                if i != mesh_dim:
+                    return False
+                else:
+                    break
```
It's possible, but I think we don't want to special treating right to left/left to right too much on the logic

wanchaol committed:
Update on "[dtensor] rewrite redistribute algorithm for multi-dim mesh"


As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D60132518](https://our.internmc.facebook.com/intern/diff/D60132518)

[ghstack-poisoned]

wanchaol commented:
> For 2D I think the algorithm is robust.
> 
> I have question on 3+ D? For example, suppose `src_placements=[S(0), S(1), S(0)]`, `dst_placements=[S(1), S(0), S(0)]`. When doing right to left we get: `[S(0), S(1), S(0)]` -> `[S(0), S(1), S(0)]` (`mesh_dim` = 2) `[S(0), S(1), S(0)]` -> `[S(0), R, S(0)]` (`mesh_dim` = 1) `[S(0), R, S(0)]` -> `[R, R, S(0)]` (`mesh_dim` = 0) But here the S(0) on `mesh_dim=2` actually becomes a strided sharding implicitly. Maybe this implicit strided sharding is OK?
> 

@tianyu-l Thanks for pointing this out! I think it's not ok to have implicit strided sharding, so I adjusted the algorithm a bit to disallow this from happening, and added unit tests to guard this .

tianyu-l commented on torch/distributed/_tensor/_redistribute.py:
```diff
@@ -97,43 +87,135 @@ def _gen_transform_infos(
+
+    # For current that is Shard, check if it's the inner most sharding
+    # if not then it's not reshardable, else we check the target sharding
+    if current.is_shard():
```
In case `[S(1), S(0), S(0)]` => `[R, S(0), S(0)]` would the following happen?

`[S(1), S(0), S(0)]` -> `[S(1), S(0), S(0)]` (`mesh_dim` = 2)
`[S(1), S(0), S(0)]` -> `[S(1), R, S(0)]` (`mesh_dim` = 1) as `reshardable_from_src_to_dst` returns `False` here
`[S(1), R, S(0)]` -> `[R, R, S(0)]` (`mesh_dim` = 0)
Again dim 2 becomes strided sharding.

Ideally there should be only one resharding at dim 0.

tianyu-l replied:
It seems to me that if we remove this check, the algorithm would be correct, although the method `reshardable_from_src_to_dst` itself won't be as general as you designed it for. In some sense I feel we don't need general check here, since we are only doing right-to-left, and left-to-right, and in left-to-right, we never need to check.

wanchaol replied:
OK i think this make sense, let me refactor the current code to make it less general, and just stick with two pass of traversal

wanchaol replied:
updated

tianyu-l commented on torch/distributed/_tensor/_redistribute.py:
```diff
@@ -97,43 +87,135 @@ def _gen_transform_infos(
+    return transform_infos
+
+
+def reshardable_from_src_to_dst(
```
shall this also be renamed as the arg names change?

tianyu-l reviewed (commented):
please see inline comments.

wanchaol committed:
Update on "[dtensor] rewrite redistribute algorithm for multi-dim mesh"


As titled, this PR rewrite the current redistribute algorithm to make
the multi-mesh dim redistribute logic more sound. The previous algorithm
works numerically but it could incur additional non-necessary steps
when transforming shardings in the multi-dimesnion device mesh, i.e.

Let's say we want to transform from (S(1), S(1)) -> (S(1), S(2)). The
previous algorithm yield the following steps:

* mesh_dim 1: S(1) -> R, mesh_dim 0: S(1) -> R
* mesh_dim 0: R -> S(1), mesh_dim 1: R -> S(2)

Although it works semantically but it incurs two allgather
transformations, where it should really only incur a S(1) -> S(2) on the
mesh dim 1.

The rewrite algorithm basically take it in a more principled way:

1. we check if src_spec have sharding, if not, we don't need to worry about nested sharding case, as sharding would always be in order, so we just go from left to right in the placements and add the transform steps
2. if src_spec have sharding, this potentially means that there would be either nested or mis-aligned shardings. So we first tranverse from right to left to check if there's mis-aligned sharding as the above example showed, if there is, we replicate that mesh dimension so that it unshard the nested sharding
3. we tranverse again from left to right to generate the transformation
   after we unshard the nested sharding

should also fix https://github.com/pytorch/pytorch/issues/132751

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o


[ghstack-poisoned]

tianyu-l reviewed (approved):
The code reads clear! Thanks!

Had a inline comment.

wanchaol commented on torch/distributed/_tensor/_redistribute.py:
```diff
@@ -87,135 +86,64 @@ def _gen_transform_infos(
-    if reshardable_from_src_to_dst(src_placements, dst_placements, current_mesh_idx):
-        # current src_placements can transform directly to dst_placements on mesh_idx
-        if src != dst:
             transform_infos.append(
```
Yeah we can keep that I feel, I'll have a follow up on this to avoid re-triggering CI

wanchaol commented:
@pytorchbot merge
